src/provider/provider_os_memory.c

```diff
@@ -0,0 +1,299 @@
+                    "bytes\n",
+                    (uintptr_t)addr, alignment, alignment);
+        }
+        goto err_unmap;
+    }
+
+    int ret = mbind(addr, size, os_params->numa_mode, os_params->nodemask,
+                    os_params->maxnode, os_params->numa_flags);
+    if (ret) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_MBIND_FAILED;
+        if (os_params->traces) {
+            perror("mbind() failed");
+        }
+        goto err_unmap;
+    }
+
+    if (os_params->traces) {
+        int numa_node = -1;
+        int ret = get_mempolicy(&numa_node, NULL, 0, (void *)addr,
+                                MPOL_F_NODE | MPOL_F_ADDR);
+        if (ret) {
+            fprintf(stderr,
+                    "cannot print assigned NUMA node - get_mempolicy() failed "
+                    "(errno = %i)\n",
+                    errno);
+            perror("get_mempolicy()");
+        } else {
+            printf("alloc(%zu) = 0x%lx, assigned NUMA node = %i\n", size,
+                   (uintptr_t)addr, numa_node);
+        }
+    }
+
+    *resultPtr = addr;
+
+    return UMF_RESULT_SUCCESS;
+
+err_unmap:
+    (void)munmap(addr, size);
+    return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+}
+
+static enum umf_result_t os_free(void *provider, void *ptr, size_t size) {
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    int ret = munmap(ptr, size);
+    if (ret) {
+        TLS_last_native_error = UMF_OS_RESULT_ERROR_MUNMAP_FAILED;
+        if (os_params->traces) {
+            perror("munmap() failed");
+        }
+        return UMF_RESULT_ERROR_MEMORY_PROVIDER_SPECIFIC;
+    }
+
+    return UMF_RESULT_SUCCESS;
+}
+
+void os_get_last_native_error(void *provider, const char **ppMessage,
+                              int32_t *pError) {
+    (void)provider; // unused
+
+    if (ppMessage == NULL || pError == NULL) {
+        assert(0);
+        return;
+    }
+
+    *pError = TLS_last_native_error;
+    *ppMessage = Native_error_str[*pError];
+}
+
+enum umf_result_t os_get_recommended_page_size(void *provider, size_t size,
+                                               size_t *pageSize) {
+    (void)size; // unused
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
+    umf_os_memory_provider_params_t *os_params = &os_provider->params;
+
+    if (pageSize == NULL) {
+        if (os_params->traces) {
+            fprintf(stderr, "the pageSize argument cannot be NULL\n");
+        }
+        assert(0);
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    *pageSize = sysconf(_SC_PAGE_SIZE);
+
+    return UMF_RESULT_SUCCESS;
+}
+
+enum umf_result_t os_get_min_page_size(void *provider, void *ptr,
+                                       size_t *pageSize) {
+    (void)ptr; // unused
+    os_memory_provider_t *os_provider = (os_memory_provider_t *)provider;
```

Comment:
please just call os_get_recommended_page_size here

Reply:
Cannot - they have different arguments.

Reply:
so what? you don't use "size" arg anyway

Reply:
OK, done